src/compose/table.c

```diff
@@ -226,3 +227,148 @@ xkb_compose_table_new_from_locale(struct xkb_context *ctx,
+    iter->table = xkb_compose_table_ref(table);
+    sequence = calloc(MAX_LHS_LEN, sizeof(xkb_keysym_t));
+    if (!sequence) {
+        free(iter);
+        return NULL;
+    }
+    iter->entry.sequence = sequence;
+    iter->entry.sequence_length = 0;
+
+    darray_init(iter->cursors);
+    cursor.direction = NODE_LEFT;
+    cursor.node_offset = 1;
+    darray_append(iter->cursors, cursor);
+
+    return iter;
+}
+
+XKB_EXPORT void
+xkb_compose_table_iterator_unref(struct xkb_compose_table_iterator *iter)
+{
+    xkb_compose_table_unref(iter->table);
+    darray_free(iter->cursors);
+    free(iter->entry.sequence);
+    free(iter);
+}
+
+XKB_EXPORT struct xkb_compose_table_entry *
+xkb_compose_table_iterator_current(struct xkb_compose_table_iterator *iter)
+{
+    return &iter->entry;
+}
+
+XKB_EXPORT struct xkb_compose_table_entry *
+xkb_compose_table_iterator_next(struct xkb_compose_table_iterator *iter)
+{
+    struct xkb_compose_table_iterator_cursor *cursor;
+    struct xkb_compose_table_iterator_cursor new_cursor;
+    const struct compose_node *node;
+
+    cursor = &darray_item(iter->cursors, iter->cursors.size - 1);
+    new_cursor.direction = NODE_LEFT;
+
+    while (cursor->direction < NODE_UP) {
+        node = &darray_item(iter->table->nodes, cursor->node_offset);
+
+        if (cursor->direction == NODE_LEFT) {
+            /* Left node */
+            cursor->direction = NODE_DOWN;
+            if (node->lokid) {
+                new_cursor.node_offset = node->lokid;
+                darray_append(iter->cursors, new_cursor);
+                cursor = &darray_item(iter->cursors, iter->cursors.size - 1);
+                continue;
+            }
+        }
+        if (cursor->direction == NODE_DOWN) {
+            /* Down node */
+            cursor->direction = NODE_RIGHT;
+            assert (iter->entry.sequence_length <= MAX_LHS_LEN);
+            iter->entry.sequence[iter->entry.sequence_length] = node->keysym;
+            iter->entry.sequence_length++;
+            if (node->is_leaf) {
+                iter->entry.keysym = node->leaf.keysym;
+                iter->entry.utf8 = &darray_item(iter->table->utf8, node->leaf.utf8);
+                return &iter->entry;
+            } else {
+                new_cursor.node_offset = node->internal.eqkid;
+                darray_append(iter->cursors, new_cursor);
+                cursor = &darray_item(iter->cursors, iter->cursors.size - 1);
+                continue;
+            }
+        }
+        cursor->direction = NODE_UP;
+        iter->entry.sequence_length--;
+        if (node->hikid) {
+            /* Right node */
+            new_cursor.node_offset = node->hikid;
+            darray_append(iter->cursors, new_cursor);
+            cursor = &darray_item(iter->cursors, iter->cursors.size - 1);
+            continue;
+        } else {
+            /* Look for next node in parents */
+            while (iter->cursors.size > 1) {
+                iter->cursors.size--;
```

Comment:
Let's add a `darray_pop` operation (or something like it) for this, should try to not break the darray encapsulation.

Reply:
I added `darray_remove_last`, as pop returns an element that we would not use.